File: LeetCode150/14.cpp
#include<bits/stdc++.h>
using namespace std;

// This is the solution of the problem Longest Common Prefix on leetcode.
// https://leetcode.com/problems/longest-common-prefix/

// The logic is very simple. We just need to find the longest common prefix of the first two strings and then use that to find the longest common prefix of the next string and so on.
// The time complexity is O(n*m) where n is the number of strings and m is the length of the longest string.
class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        int n = strs.size();
        if(n == 0)
            return "";
        string ans = strs[0];
        for(int i = 1; i < n; i++)
        {
            int j = 0;
            while(j < ans.length() && j < strs[i].length() && ans[j] == strs[i][j])
                j++;
            ans = ans.substr(0, j);
        }
        return ans;
    }
};
int main()
{
    return 0;
}

File: LeetCode150/14.py
class Solution:
    def findCommon(self, s1, s2):
        for i in range(min(len(s1), len(s2))):
            if s1[i] != s2[i]:
                return s1[:i]
            
        return s1 if len(s1) <= len(s2) else s2

    def longestCommonPrefix(self, strs: list[str]) -> str:
        common = strs[0]
        for i in range(1, len(strs)):
            common = self.findCommon(common, strs[i])
        return common

s = Solution()
strs = ["abcd", "abc"]
common = s.longestCommonPrefix(strs)
print(common)